feat: Prefix Command Management

pdellaert opened the pull request:
# Prefix Command Management

## Description

Please read https://github.com/pdellaert/fbw-discord-bot-utils/blob/dynamic-command-management/docs/prefix-commands.md for all the fancy details.

NOTE: The cache refresh system will be rewritten to be safer, right now it isn't perfectly safe. 

## Test Results

I don't know how many hours, but probably good someone takes their own look... 

## Discord Username

straks

pdellaert committed:
Initial prefix commands management

pdellaert committed:
Splitting into multiple commands

pdellaert committed:
Migrating Schema and adapting Content management

pdellaert committed:
Updating permissions for new schema and fix issues

pdellaert committed:
Adding AutoComplete and Content Modal

* AutoComplete added for categories, versions and command names where
  appropriate
* Added Modal for Content

pdellaert committed:
Adding Default Channel Versions and add autocomplete to everything

pdellaert committed:
Introducing cache mechanism to keep local cache, manage it and initial message create handler

pdellaert committed:
Making cache and command checks case-insensitive

pdellaert committed:
Adding a scheduled task to refresh the cache automatically

pdellaert committed:
Adding Categories and Default Channel Versions to cache

pdellaert committed:
Managing Category cache and Default Version Cache

pdellaert committed:
Adding default channel version parsing

pdellaert committed:
Refactor cache management to be more intuitive and introduce version selection for GENERIC versions

pdellaert committed:
Fixing staging config json

pdellaert committed:
Refactor Permission mechanism

pdellaert committed:
Finalizing permission work

pdellaert committed:
Adding description to command and fixing autoComplete for Prefix commands

pdellaert committed:
Made changes to Prefix Command Handler and Cache Manager:
* Cache Manager: Moved cache prefixes to variables and export them
* Handler: Use new cache prefix variables
* Handler: Removed unnecessary footer checks (prefix commands have no
  footer)
* Handler: Introduced expiration edit if version not selected in time
* Handler: If a specific version is asked, but doesn't exist generic is
  given without options
* Handler: Versions are now sorted to be consistent

pdellaert committed:
Introducing prefix help to show list of prefix commands per category

pdellaert committed:
Use appropriate language for blocklist

pdellaert committed:
Making the max length more standardized

pdellaert committed:
Show the emoji of a category

pdellaert committed:
Show the buttons for generic in case the version specific content does not exist

pdellaert committed:
Fixing bug when there is no CDV set for a channel

pdellaert committed:
Adding initial documentation for prefix-commands

pdellaert committed:
Refactor cache management for Channel Default Versions

pdellaert committed:
Refactoring cache

pdellaert committed:
Updating docs on cachi auto refresh

pdellaert committed:
Adding CHANGELOG entry

pdellaert committed:
Disabling autocreate for unnecessary collections

pdellaert committed:
Fixing bug where initial permissions were not properly set

pdellaert committed:
Fixing issues:
* Allowing empty title
* Fixing messageCreateHandler in case of unset permissions on command
*

pdellaert committed:
Fixing maximum number of callback items

pdellaert committed:
Fix if there's no content at all

pdellaert committed:
Fix boolean handling

benw202 commented on src/commands/moderation/prefixCommands/functions/addChannelPermission.ts:
```diff
@@ -0,0 +1,112 @@
+});
+
+const failedEmbed = (command: string, channel: string) => makeEmbed({
+    title: 'Prefix Commands - Add Channel Permission - Failed',
```
```suggestion
    title: 'Prefix Commands - Add Channel - Failed',
```

The rest just say 'Add Channel'

benw202 commented on src/commands/moderation/prefixCommands/functions/addCommand.ts:
```diff
@@ -0,0 +1,160 @@
+    const embedColor = interaction.options.getString('embed_color') || '';
+    const moderator = interaction.user;
+
+    const nameRegex = /^[\w\d-_]+$/;
```
```suggestion
    const nameRegex = /^[\w-_]+$/;
```

Is the \d redundant with the \w

benw202 commented on src/commands/moderation/prefixCommands/functions/addVersion.ts:
```diff
@@ -0,0 +1,127 @@
+    const enabled = interaction.options.getBoolean('is_enabled') || false;
+    const moderator = interaction.user;
+
+    const nameRegex = /^[\w\d-_]+$/;
```
```suggestion
    const nameRegex = /^[\w-_]+$/;
```

benw202 commented on src/commands/moderation/prefixCommands/functions/modifyCommand.ts:
```diff
@@ -0,0 +1,153 @@
+    const embedColor = interaction.options.getString('embed_color') || '';
+    const moderator = interaction.user;
+
+    const nameRegex = /^[\w\d-_]+$/;
```
```suggestion
    const nameRegex = /^[\w-_]+$/;
```

benw202 commented on src/commands/moderation/prefixCommands/functions/modifyVersion.ts:
```diff
@@ -0,0 +1,128 @@
+    const enabled = interaction.options.getBoolean('is_enabled');
+    const moderator = interaction.user;
+
+    const nameRegex = /^[\w\d-_]+$/;
```
```suggestion
    const nameRegex = /^[\w-_]+$/;
```

benw202 commented on src/commands/moderation/prefixCommands/functions/setCommandPermissionSettings.ts:
```diff
@@ -0,0 +1,123 @@
+});
+
+const modLogEmbed = (moderator: User, command: string, rolesBlocklist: boolean, channelsBlocklist: boolean, quietErrors: boolean, verboseErrors: boolean) => makeEmbed({
+    title: 'Prefix command version added',
```
Is this the wrong title?

benw202 commented on src/commands/moderation/prefixCommands/functions/addChannelPermission.ts:
```diff
@@ -0,0 +1,112 @@
+});
+
+const successEmbed = (command: string, channel: string) => makeEmbed({
+    title: `Prefix command channel <#${channel}> added for command ${command}.}`,
```
```suggestion
    title: `Prefix command channel <#${channel}> added for command ${command}.`,
```

benw202 commented on src/commands/moderation/prefixCommands/functions/addRolePermission.ts:
```diff
@@ -0,0 +1,112 @@
+});
+
+const successEmbed = (command: string, roleName: string) => makeEmbed({
+    title: `Prefix command role ${roleName} added for command ${command}.}`,
```
```suggestion
    title: `Prefix command role ${roleName} added for command ${command}.`,
```

benw202 commented on src/commands/utils/prefixHelp.ts:
```diff
@@ -0,0 +1,155 @@
+import { ApplicationCommandOptionChoiceData, ApplicationCommandOptionType, ApplicationCommandType } from 'discord.js';
+import { makeEmbed, createPaginatedEmbedHandler, slashCommand, slashCommandStructure, getInMemoryCache, memoryCachePrefixCommand, AutocompleteCallback, memoryCachePrefixCategory, makeLines, memoryCachePrefixVersion, Logger } from '../../lib';
```
```suggestion
import { makeEmbed, createPaginatedEmbedHandler, slashCommand, slashCommandStructure, getInMemoryCache, memoryCachePrefixCommand, AutocompleteCallback, memoryCachePrefixCategory, makeLines, memoryCachePrefixVersion, Logger, PrefixCommand, PrefixCommandVersion, PrefixCommandCategory, IPrefixCommand } from '../../lib';
```

Can reduce amount of imports

benw202 commented on src/commands/utils/prefixHelp.ts:
```diff
@@ -0,0 +1,155 @@
+import { ApplicationCommandOptionChoiceData, ApplicationCommandOptionType, ApplicationCommandType } from 'discord.js';
+import { makeEmbed, createPaginatedEmbedHandler, slashCommand, slashCommandStructure, getInMemoryCache, memoryCachePrefixCommand, AutocompleteCallback, memoryCachePrefixCategory, makeLines, memoryCachePrefixVersion, Logger } from '../../lib';
+import { PrefixCommand, PrefixCommandVersion, PrefixCommandCategory, IPrefixCommand } from '../../lib/schemas/prefixCommandSchemas';
```
```suggestion
```

benw202 commented on src/events/messageCreateHandler.ts:
```diff
@@ -0,0 +1,277 @@
+import { ActionRowBuilder, ButtonBuilder, ButtonInteraction, ButtonStyle, EmbedBuilder, Interaction, Message } from 'discord.js';
+import { event, getInMemoryCache, memoryCachePrefixCommand, memoryCachePrefixVersion, memoryCachePrefixChannelDefaultVersion, Logger, Events, constantsConfig, makeEmbed, makeLines } from '../lib';
```
```suggestion
import { event, getInMemoryCache, memoryCachePrefixCommand, memoryCachePrefixVersion, memoryCachePrefixChannelDefaultVersion, Logger, Events, constantsConfig, makeEmbed, makeLines, PrefixCommand, PrefixCommandPermissions, PrefixCommandVersion } from '../lib';
```

Can reduce amount of imports

benw202 commented on src/events/messageCreateHandler.ts:
```diff
@@ -0,0 +1,277 @@
+import { ActionRowBuilder, ButtonBuilder, ButtonInteraction, ButtonStyle, EmbedBuilder, Interaction, Message } from 'discord.js';
+import { event, getInMemoryCache, memoryCachePrefixCommand, memoryCachePrefixVersion, memoryCachePrefixChannelDefaultVersion, Logger, Events, constantsConfig, makeEmbed, makeLines } from '../lib';
+import { PrefixCommand, PrefixCommandPermissions, PrefixCommandVersion } from '../lib/schemas/prefixCommandSchemas';
```
```suggestion
```

benw202 commented on src/events/messageCreateHandler.ts:
```diff
@@ -0,0 +1,277 @@
+                    collector.on('collect', async (collectedInteraction: ButtonInteraction) => {
+                        Logger.debug(`Prefix Command - User selected version "${collectedInteraction.customId}" for command "${name}" based on user command "${commandText}"`);
+                        await collectedInteraction.deferUpdate();
+                        buttonMessage.delete();
```
```suggestion
                        await buttonMessage.delete();
```
Await?

benw202 reviewed (changes requested):
Really good work! I like the way this has been done.

benw202 commented:
Some further comments as discussed:

- When modifying a command name, the mod log embed doesn't show the command description
- MongoDB __V not incrementing on an update - intentional?
- On timeout of modal it says "You did not provide the necessary content information and the change was not made.". The user may have provided info, but not in time. Is it worth saying that there is a two minute timeout somewhere in the modal description?

benw202 commented:
Moderator in `Prefix Command Cache Update` is not a clickable tag. Could make this clickable if we want it to match other mod logs?

pdellaert committed:
Fix: Cache update with additional caches and proper mod output

pdellaert committed:
Fix: Improvements based on review feedback

pdellaert committed:
Fix: review issues and bug on missing max number of choices

pdellaert committed:
Change order of cache update to be in line with command

pdellaert commented:
Addressed the review comments, will look at
* MongoDB __V not incrementing on an update - intentional?
* On timeout of modal it says "You did not provide the necessary content information and the change was not made.". The user may have provided info, but not in time. Is it worth saying that there is a two minute timeout somewhere in the modal description?

Also likely to hide the choice buttons if it is a "fallback-to-GENERIC" if a different unavailable version is requested (either explicit, or by channel default)

pdellaert committed:
Fix: Fixing a potential race condition to set the clicked immediately

pdellaert committed:
Fix: Actually fixes the race condition properly by checking for the right event

pdellaert committed:
Disabling buttons if it is a fallback to GENERIC version

pdellaert commented:
> * On timeout of modal it says "You did not provide the necessary content information and the change was not made.". The user may have provided info, but not in time. Is it worth saying that there is a two minute timeout somewhere in the modal description?

Can this be done? I don't see a way to add a description to the Modal? Doesn't look like we do it anywhere?

benw202 commented:
> > * On timeout of modal it says "You did not provide the necessary content information and the change was not made.". The user may have provided info, but not in time. Is it worth saying that there is a two minute timeout somewhere in the modal description?
> 
> Can this be done? I don't see a way to add a description to the Modal? Doesn't look like we do it anywhere?

You're right! I saw descriptions somewhere but seems it was only a concept. I guess we will just have to remember that we have a two minute limit. Make sure that people who can add and modify commands write their commands out elsewhere and just copy/paste in.

ExampleWasTaken commented:
I just gave this a test and thought I'd share what I found. I have not reviewed the code yet because I wanted to understand the functionality first.

All was tested on [e95eae2](https://github.com/flybywiresim/discord-bot-utils/pull/85/commits/e95eae2ab3ceef2e7048a88f53b9433d011d613e).

To get the important stuff right out of the way:
- The bot crashes when trying to set content for an invalid/non-existing version

The rest are just thoughts I had while messing about.

## When a custom version is set as a channel default and that version is disabled, nothing is shown.
Example setup with `a32nx` and `a380x` being custom versions of the `.bg` command:

#### `a380x` is disabled but set as default version for channel A:
Running `.bg` results in nothing being shown. IMO this should result in the GENERIC version being shown. The presence of the `a380x` is not leaked as showing the GENERIC version is the default behavior anyway. Additionally, the way it currently works, there is not way to get a list of all available versions.

## Commands without content?
Is there any reason why commands can be "empty" i.e. have no content set? IMO there should at least be a GENERIC version of every command.
This will
1. ensure that users are able to retrieve a list of all available versions
2. prevents "ghost" commands that exist but have no content and therefore no use

I might be missing something here but I didn't find an explanation for this and thought I'd raise it. :)

## Consistent slash command
As it stands right now there are three different slash-commands for this feature:
- `/prefix-commands`
- `/prefix-command-permissions`
- `/prefix-help`
- `/prefix-commands-cache-update`

I think this is pretty inconsistent and confusing to remember (was it `/prefix-commands-permissions`, `/prefix-command-permission` or `/prefix-command-permissions`?). 
Instead, I suggest we use a common slash-command `/prefix` and have everything else be sub commands.

Example:

- `/prefix categories <...>`
- `/prefix versions <...>`
- `/prefix permissions <...>`
- `/prefix help <...>`
- `/prefix cache <...>`

This way, the user can explore all available commands in the Discord client by just typing `/prefix `. It also reduces the overall slash-commands used and provides a cleaner command structure imo.

---

I think overall this is looking pretty awesome! Really nice work!

ExampleWasTaken commented:
Another bug I just found:

Aliases are not unique against command names and aliases of other commands.

Example:

These can all exist at the same time.

## Command 1
- Name: `beginners_guide`
- Alias:   `bg`

## Command 2
- Name: `bg` 

## Command 3
- Name: `test`
- Alias:   `bg`

---

This results in undefined behavior as there are three possible commands that would all fulfil the conditions to be shown.

pdellaert committed:
Fix bugs - crash for unexisting version in setContext and missing db returns

pdellaert committed:
Fix bugs:
* Showing buttons for disabled versions on generic content
* Not showing anything if channel default is disabled

pdellaert committed:
Fix bug: Add name and aliases uniqueness

pdellaert committed:
Fix bugs in Content management
* Not properly checking commands and versions
* Not prorperly using defer/reply/followUp

pdellaert committed:
Fix: Checking for empty content

pdellaert committed:
Fix verifying generic does not get created/used for versions or commands. And version aliases do not overlap with commands

pdellaert committed:
Fix bug where if no content for enabled versions was found it could crash

pdellaert committed:
Clarifying button

pdellaert commented:
Bunch of changes, most are bug fixes and minor enhancements to make sure that there's no overlap in commands, aliases & versions. Also no empty contents. 

Feedback on commands without content: Can't really do anything about it. I can't force setting a GENERIC content, unless we do it during the creation of the command, but would require a bit more work than we have time for right now.

Feeback on different slash commands: As discussed in the channel, Discord has a max length on command definition and subgroup/commands, we would be exceeding it if we put it in one. 

Changes since my last comment (4 days ago) https://github.com/pdellaert/fbw-discord-bot-utils/compare/e95eae2ab3ceef2e7048a88f53b9433d011d613e...dynamic-command-management

ExampleWasTaken commented:
That sounds awesome! 

Regarding the commands without content, I think that something we can introduce down the line if we want to. Definitely not a showstopper.

benw202 commented:
> That sounds awesome!
> 
> Regarding the commands without content, I think that something we can introduce down the line if we want to. Definitely not a showstopper.

Agree on this. There will be a limited amount of people who can add commands so the likelihood of us having rogue empty commands should be none.

ExampleWasTaken commented on src/commands/moderation/prefixCommands/functions/addCategory.ts:
```diff
@@ -0,0 +1,98 @@
+
+    //Check if the mod logs channel exists
+    const modLogsChannel = interaction.guild.channels.resolve(constantsConfig.channels.MOD_LOGS) as TextChannel;
+    if (!modLogsChannel) {
```
```suggestion
    let modLogsChannel = interaction.guild.channels.resolve(constantsConfig.channels.MOD_LOGS);
    if (!modLogsChannel || !modLogsChannel.isTextBased()) {
        modLogsChannel = null;
```
Might be safer to actually check if the channel is a text channel? Especially if we want to offer the codebase to other projects. Then, if it's not a text channel set it to null to prevent accidental attempts to send into a non-text channel.

ExampleWasTaken commented on src/commands/moderation/prefixCommands/functions/modifyCategory.ts:
```diff
@@ -0,0 +1,99 @@
+
+const doesNotExistsEmbed = (category: string) => makeEmbed({
+    title: 'Prefix Commands - Modify Category - Does not exist',
+    description: `The prefix command category ${category} does not exists. Can not modify it.`,
```
```suggestion
    description: `The prefix command category ${category} does not exists. Cannot modify it.`,
```
"Cannot" is generally more common. I suggest we use that but obviously that's personal preference and I'm happy to leave it as is!

ExampleWasTaken commented on src/commands/moderation/prefixCommands/functions/modifyCategory.ts:
```diff
@@ -0,0 +1,99 @@
+
+    //Check if the mod logs channel exists
+    const modLogsChannel = interaction.guild.channels.resolve(constantsConfig.channels.MOD_LOGS) as TextChannel;
+    if (!modLogsChannel) {
```
```suggestion
    let modLogsChannel = interaction.guild.channels.resolve(constantsConfig.channels.MOD_LOGS);
    if (!modLogsChannel || !modLogsChannel.isTextBased()) {
        modLogsChannel = null;
```
Might be safer to actually check if the channel is a text channel? Especially if we want to offer the codebase to other projects. Then, if it's not a text channel set it to null to prevent accidental attempts to send into a non-text channel.

ExampleWasTaken commented on src/commands/moderation/prefixCommands/functions/deleteCategory.ts:
```diff
@@ -0,0 +1,93 @@
+
+const doesNotExistsEmbed = (category: string) => makeEmbed({
+    title: 'Prefix Commands - Delete Category - Does not exist',
+    description: `The prefix command category ${category} does not exists. Can not delete it.`,
```
```suggestion
    description: `The prefix command category ${category} does not exists. Cannot delete it.`,
```
"Cannot" is generally more common. I suggest we use that but obviously that's personal preference and I'm happy to leave it as is!

ExampleWasTaken commented on src/commands/moderation/prefixCommands/functions/addCategory.ts:
```diff
@@ -0,0 +1,98 @@
+    const modLogsChannel = interaction.guild.channels.resolve(constantsConfig.channels.MOD_LOGS) as TextChannel;
+    if (!modLogsChannel) {
+        await interaction.followUp({ embeds: [noModLogs], ephemeral: true });
+        return;
```
```suggestion
        await interaction.followUp({ embeds: [noModLogs], ephemeral: true });
```
As discussed, this is just a warning and shouldn't block the creation.

ExampleWasTaken commented on src/commands/moderation/prefixCommands/functions/deleteCategory.ts:
```diff
@@ -0,0 +1,93 @@
+
+    //Check if the mod logs channel exists
+    const modLogsChannel = interaction.guild.channels.resolve(constantsConfig.channels.MOD_LOGS) as TextChannel;
+    if (!modLogsChannel) {
```
```suggestion
    let modLogsChannel = interaction.guild.channels.resolve(constantsConfig.channels.MOD_LOGS);
    if (!modLogsChannel || !modLogsChannel.isTextBased()) {
        modLogsChannel = null;
```
Might be safer to actually check if the channel is a text channel? Especially if we want to offer the codebase to other projects. Then, if it's not a text channel set it to null to prevent accidental attempts to send into a non-text channel.

ExampleWasTaken commented on src/commands/moderation/prefixCommands/functions/addVersion.ts:
```diff
@@ -0,0 +1,142 @@
+    }
+    if (!nameRegex.test(alias)) {
+        // eslint-disable-next-line no-await-in-loop
+        await interaction.followUp({ embeds: [wrongFormatEmbed(alias)], ephemeral: true });
```
```suggestion
    if (!nameRegex.test(alias)) {
        await interaction.followUp({ embeds: [wrongFormatEmbed(alias)], ephemeral: true });
```
I don't believe that ESLint directive is needed here?

ExampleWasTaken commented on src/commands/moderation/prefixCommands/functions/addVersion.ts:
```diff
@@ -0,0 +1,142 @@
+
+    //Check if the mod logs channel exists
+    const modLogsChannel = interaction.guild.channels.resolve(constantsConfig.channels.MOD_LOGS) as TextChannel;
+    if (!modLogsChannel) {
```
```suggestion
    let modLogsChannel = interaction.guild.channels.resolve(constantsConfig.channels.MOD_LOGS);
    if (!modLogsChannel || !modLogsChannel.isTextBased()) {
        modLogsChannel = null;
```
Might be safer to actually check if the channel is a text channel? Especially if we want to offer the codebase to other projects. Then, if it's not a text channel set it to null to prevent accidental attempts to send into a non-text channel.

ExampleWasTaken commented on src/commands/moderation/prefixCommands/functions/modifyVersion.ts:
```diff
@@ -0,0 +1,166 @@
+
+const doesNotExistsEmbed = (version: string) => makeEmbed({
+    title: 'Prefix Commands - Modify Version - Does not exist',
+    description: `The prefix command version ${version} does not exists. Can not modify it.`,
```
```suggestion
    description: `The prefix command version ${version} does not exists. Cannot modify it.`,
```
"Cannot" is generally more common. I suggest we use that but obviously that's personal preference and I'm happy to leave it as is!

ExampleWasTaken commented on src/commands/moderation/prefixCommands/functions/modifyVersion.ts:
```diff
@@ -0,0 +1,166 @@
+    }
+    if (alias && !nameRegex.test(alias)) {
+        // eslint-disable-next-line no-await-in-loop
+        await interaction.followUp({ embeds: [wrongFormatEmbed(alias)], ephemeral: true });
```
```suggestion
    if (alias && !nameRegex.test(alias)) {
        await interaction.followUp({ embeds: [wrongFormatEmbed(alias)], ephemeral: true });
```
I don't believe that ESLint directive is needed here?

ExampleWasTaken commented on src/commands/moderation/prefixCommands/functions/modifyVersion.ts:
```diff
@@ -0,0 +1,166 @@
+
+    //Check if the mod logs channel exists
+    const modLogsChannel = interaction.guild.channels.resolve(constantsConfig.channels.MOD_LOGS) as TextChannel;
+    if (!modLogsChannel) {
```
```suggestion
    let modLogsChannel = interaction.guild.channels.resolve(constantsConfig.channels.MOD_LOGS);
    if (!modLogsChannel || !modLogsChannel.isTextBased()) {
        modLogsChannel = null;
```
Might be safer to actually check if the channel is a text channel? Especially if we want to offer the codebase to other projects. Then, if it's not a text channel set it to null to prevent accidental attempts to send into a non-text channel.

ExampleWasTaken commented on src/commands/moderation/prefixCommands/functions/deleteVersion.ts:
```diff
@@ -0,0 +1,149 @@
+
+const doesNotExistsEmbed = (version: string) => makeEmbed({
+    title: 'Prefix Commands - Delete Version - Does not exist',
+    description: `The prefix command version ${version} does not exists. Can not delete it.`,
```
```suggestion
    description: `The prefix command version ${version} does not exists. Cannot delete it.`,
```
"Cannot" is generally more common. I suggest we use that but obviously that's personal preference and I'm happy to leave it as is!

ExampleWasTaken commented on src/commands/moderation/prefixCommands/functions/deleteVersion.ts:
```diff
@@ -0,0 +1,149 @@
+    //Check if the mod logs channel exists
+    const modLogsChannel = interaction.guild.channels.resolve(constantsConfig.channels.MOD_LOGS) as TextChannel;
+    if (!modLogsChannel) {
+        await interaction.followUp({ embeds: [noModLogs], ephemeral: true });
```
```suggestion
    let modLogsChannel = interaction.guild.channels.resolve(constantsConfig.channels.MOD_LOGS);
    if (!modLogsChannel || !modLogsChannel.isTextBased()) {
        modLogsChannel = null;
```
Might be safer to actually check if the channel is a text channel? Especially if we want to offer the codebase to other projects. Then, if it's not a text channel set it to null to prevent accidental attempts to send into a non-text channel.

ExampleWasTaken commented on src/commands/moderation/prefixCommands/functions/addCommand.ts:
```diff
@@ -0,0 +1,178 @@
+
+    //Check if the mod logs channel exists
+    const modLogsChannel = interaction.guild.channels.resolve(constantsConfig.channels.MOD_LOGS) as TextChannel;
+    if (!modLogsChannel) {
```
```suggestion
    let modLogsChannel = interaction.guild.channels.resolve(constantsConfig.channels.MOD_LOGS);
    if (!modLogsChannel || !modLogsChannel.isTextBased()) {
        modLogsChannel = null;
```
Might be safer to actually check if the channel is a text channel? Especially if we want to offer the codebase to other projects. Then, if it's not a text channel set it to null to prevent accidental attempts to send into a non-text channel.

ExampleWasTaken commented on src/commands/moderation/prefixCommands/functions/modifyCommand.ts:
```diff
@@ -0,0 +1,205 @@
+
+const doesNotExistsEmbed = (command: string) => makeEmbed({
+    title: 'Prefix Commands - Modify Command - Does not exist',
+    description: `The prefix command ${command} does not exists. Can not modify it.`,
```
```suggestion
    description: `The prefix command ${command} does not exists. Cannot modify it.`,
```
"Cannot" is generally more common. I suggest we use that but obviously that's personal preference and I'm happy to leave it as is!

ExampleWasTaken commented on src/commands/moderation/prefixCommands/functions/modifyCommand.ts:
```diff
@@ -0,0 +1,205 @@
+
+    //Check if the mod logs channel exists
+    const modLogsChannel = interaction.guild.channels.resolve(constantsConfig.channels.MOD_LOGS) as TextChannel;
+    if (!modLogsChannel) {
```
```suggestion
    let modLogsChannel = interaction.guild.channels.resolve(constantsConfig.channels.MOD_LOGS);
    if (!modLogsChannel || !modLogsChannel.isTextBased()) {
        modLogsChannel = null;
```
Might be safer to actually check if the channel is a text channel? Especially if we want to offer the codebase to other projects. Then, if it's not a text channel set it to null to prevent accidental attempts to send into a non-text channel.

ExampleWasTaken commented on src/commands/moderation/prefixCommands/functions/deleteCommand.ts:
```diff
@@ -0,0 +1,105 @@
+
+const doesNotExistsEmbed = (command: string) => makeEmbed({
+    title: 'Prefix Commands - Delete Command - Does not exist',
+    description: `The prefix command ${command} does not exists. Can not delete it.`,
```
```suggestion
    description: `The prefix command ${command} does not exists. Cannot delete it.`,
```
"Cannot" is generally more common. I suggest we use that but obviously that's personal preference and I'm happy to leave it as is!

ExampleWasTaken commented on src/commands/moderation/prefixCommands/functions/deleteCommand.ts:
```diff
@@ -0,0 +1,105 @@
+
+    //Check if the mod logs channel exists
+    const modLogsChannel = interaction.guild.channels.resolve(constantsConfig.channels.MOD_LOGS) as TextChannel;
+    if (!modLogsChannel) {
```
```suggestion
    let modLogsChannel = interaction.guild.channels.resolve(constantsConfig.channels.MOD_LOGS);
    if (!modLogsChannel || !modLogsChannel.isTextBased()) {
        modLogsChannel = null;
```
Might be safer to actually check if the channel is a text channel? Especially if we want to offer the codebase to other projects. Then, if it's not a text channel set it to null to prevent accidental attempts to send into a non-text channel.

ExampleWasTaken commented on src/commands/moderation/prefixCommands/functions/setContent.ts:
```diff
@@ -0,0 +1,232 @@
+        .setLabel('Image URL')
+        .setPlaceholder('Provide an optional Image URL for the command.')
+        .setStyle(TextInputStyle.Short)
+        .setMaxLength(255)
```
Does it make sense to limit the length of the URL? URLs quickly get longer than 255 characters.

ExampleWasTaken commented on src/commands/moderation/prefixCommands/functions/setContent.ts:
```diff
@@ -0,0 +1,232 @@
+
+    await interaction.showModal(contentModal);
+
+    const filter = (interaction: {
+        customId: string;
+        user: { id: any; };
+    }) => interaction.customId === 'commandContentModal' && interaction.user.id;
```
```suggestion
    const filter = (interaction: ModalSubmitInteraction) => interaction.customId === 'commandContentModal' && interaction.user.id;
```
Why not use the actual type?

pdellaert replied:
Good suggestion, and it actually indicated a bug! I intended t check that the interaction user was actually the user who executed the command in the first place, which was missing!

ExampleWasTaken commented on src/commands/moderation/prefixCommands/functions/setContent.ts:
```diff
@@ -0,0 +1,232 @@
+        //Handle the error if the user does not respond in time
+        Logger.error(error);
+        await interaction.followUp({
+            content: 'You did not provide the necessary content information and the change was not made.',
```
```suggestion
            content: 'You did not provide the necessary content information in time (2 minutes) and the change was not made.',
```
I would include the information that there is a 2-minute limit in the error message. That way, users get educated about the limitation should they ever come across it.

ExampleWasTaken commented on src/commands/moderation/prefixCommands/functions/setContent.ts:
```diff
@@ -0,0 +1,232 @@
+    }
+    //Check if the mod logs channel exists
+    const modLogsChannel = interaction.guild.channels.resolve(constantsConfig.channels.MOD_LOGS) as TextChannel;
+    if (!modLogsChannel) {
```
```suggestion
    let modLogsChannel = interaction.guild.channels.resolve(constantsConfig.channels.MOD_LOGS);
    if (!modLogsChannel || !modLogsChannel.isTextBased()) {
        modLogsChannel = null;
```
Might be safer to actually check if the channel is a text channel? Especially if we want to offer the codebase to other projects. Then, if it's not a text channel set it to null to prevent accidental attempts to send into a non-text channel.

ExampleWasTaken commented on src/commands/moderation/prefixCommands/functions/deleteContent.ts:
```diff
@@ -0,0 +1,145 @@
+
+    //Check if the mod logs channel exists
+    const modLogsChannel = interaction.guild.channels.resolve(constantsConfig.channels.MOD_LOGS) as TextChannel;
+    if (!modLogsChannel) {
```
```suggestion
    let modLogsChannel = interaction.guild.channels.resolve(constantsConfig.channels.MOD_LOGS);
    if (!modLogsChannel || !modLogsChannel.isTextBased()) {
        modLogsChannel = null;
```
Might be safer to actually check if the channel is a text channel? Especially if we want to offer the codebase to other projects. Then, if it's not a text channel set it to null to prevent accidental attempts to send into a non-text channel.

pdellaert committed:
Fix modchannel checks

ExampleWasTaken commented on src/commands/moderation/prefixCommands/functions/setChannelDefaultVersion.ts:
```diff
@@ -0,0 +1,108 @@
+
+    //Check if the mod logs channel exists
+    const modLogsChannel = interaction.guild.channels.resolve(constantsConfig.channels.MOD_LOGS) as TextChannel;
+    if (!modLogsChannel) {
```
```suggestion
    let modLogsChannel = interaction.guild.channels.resolve(constantsConfig.channels.MOD_LOGS);
    if (!modLogsChannel || !modLogsChannel.isTextBased()) {
        modLogsChannel = null;
```
Might be safer to actually check if the channel is a text channel? Especially if we want to offer the codebase to other projects. Then, if it's not a text channel set it to null to prevent accidental attempts to send into a non-text channel.

ExampleWasTaken commented on src/commands/moderation/prefixCommands/functions/deleteChannelDefaultVersion.ts:
```diff
@@ -0,0 +1,88 @@
+
+    //Check if the mod logs channel exists
+    const modLogsChannel = interaction.guild.channels.resolve(constantsConfig.channels.MOD_LOGS) as TextChannel;
+    if (!modLogsChannel) {
```
```suggestion
    let modLogsChannel = interaction.guild.channels.resolve(constantsConfig.channels.MOD_LOGS);
    if (!modLogsChannel || !modLogsChannel.isTextBased()) {
        modLogsChannel = null;
```
Might be safer to actually check if the channel is a text channel? Especially if we want to offer the codebase to other projects. Then, if it's not a text channel set it to null to prevent accidental attempts to send into a non-text channel.

ExampleWasTaken commented on src/commands/moderation/prefixCommands/functions/addChannelPermission.ts:
```diff
@@ -0,0 +1,112 @@
+
+    //Check if the mod logs channel exists
+    const modLogsChannel = interaction.guild.channels.resolve(constantsConfig.channels.MOD_LOGS) as TextChannel;
+    if (!modLogsChannel) {
```
```suggestion
    let modLogsChannel = interaction.guild.channels.resolve(constantsConfig.channels.MOD_LOGS);
    if (!modLogsChannel || !modLogsChannel.isTextBased()) {
        modLogsChannel = null;
```
Might be safer to actually check if the channel is a text channel? Especially if we want to offer the codebase to other projects. Then, if it's not a text channel set it to null to prevent accidental attempts to send into a non-text channel.

ExampleWasTaken commented on src/commands/moderation/prefixCommands/functions/addRolePermission.ts:
```diff
@@ -0,0 +1,112 @@
+    const moderator = interaction.user;
+
+    //Check if the mod logs role exists
+    const modLogsRole = interaction.guild.channels.resolve(constantsConfig.channels.MOD_LOGS) as TextChannel;
```
I think this is meant to be called `modLogsChannel`?

ExampleWasTaken commented on src/commands/moderation/prefixCommands/functions/removeChannelPermission.ts:
```diff
@@ -0,0 +1,109 @@
+
+    //Check if the mod logs channel exists
+    const modLogsChannel = interaction.guild.channels.resolve(constantsConfig.channels.MOD_LOGS) as TextChannel;
+    if (!modLogsChannel) {
```
```suggestion
    let modLogsChannel = interaction.guild.channels.resolve(constantsConfig.channels.MOD_LOGS);
    if (!modLogsChannel || !modLogsChannel.isTextBased()) {
        modLogsChannel = null;
```
Might be safer to actually check if the channel is a text channel? Especially if we want to offer the codebase to other projects. Then, if it's not a text channel set it to null to prevent accidental attempts to send into a non-text channel.

ExampleWasTaken commented on src/commands/moderation/prefixCommands/functions/removeRolePermission.ts:
```diff
@@ -0,0 +1,109 @@
+    const moderator = interaction.user;
+
+    //Check if the mod logs role exists
+    const modLogsRole = interaction.guild.channels.resolve(constantsConfig.channels.MOD_LOGS) as TextChannel;
```
I think this is meant to be called `modLogsChannel`?

ExampleWasTaken commented on src/commands/moderation/prefixCommands/functions/setCommandPermissionSettings.ts:
```diff
@@ -0,0 +1,122 @@
+
+    //Check if the mod logs channel exists
+    const modLogsChannel = interaction.guild.channels.resolve(constantsConfig.channels.MOD_LOGS) as TextChannel;
+    if (!modLogsChannel) {
```
```suggestion
    let modLogsChannel = interaction.guild.channels.resolve(constantsConfig.channels.MOD_LOGS);
    if (!modLogsChannel || !modLogsChannel.isTextBased()) {
        modLogsChannel = null;
```
Might be safer to actually check if the channel is a text channel? Especially if we want to offer the codebase to other projects. Then, if it's not a text channel set it to null to prevent accidental attempts to send into a non-text channel.

ExampleWasTaken commented on src/commands/moderation/prefixCommands/prefixCommandPermissions.ts:
```diff
@@ -0,0 +1,249 @@
+    const name = Object.keys(constantsConfig.colors)[i];
+    const value = constantsConfig.colors[name];
+    colorChoices.push({ name, value });
+}
```
```suggestion
```
This is never used. Probably a copy/paste left-over from `prefixCommands.ts`.

ExampleWasTaken commented on src/commands/moderation/prefixCommands/prefixCommandCacheUpdate.ts:
```diff
@@ -0,0 +1,74 @@
+    options: [],
+});
+
+const cacheUpdateEmbed = (fields: any, color: number) => makeEmbed({
```
```suggestion
const cacheUpdateEmbed = (fields: APIEmbedField[], color: number) => makeEmbed({
```
Why not use the actual type? It has to be `APIEmbedField[]` anyway.

ExampleWasTaken commented on src/events/messageCreateHandler.ts:
```diff
@@ -0,0 +1,298 @@
+            const commandCachedVersion = await inMemoryCache.get(`${memoryCachePrefixVersion}:${commandText.toLowerCase()}`);
+            let commandVersionId;
+            let commandVersionName;
+            let commandVersionEnabled;
```
```suggestion
            let commandVersionId: string;
            let commandVersionName: string;
            let commandVersionEnabled: boolean;
```
I think this should be possible? I would prefer this as it gets rid of `any`.

ExampleWasTaken commented on src/lib/cache/cacheManager.ts:
```diff
@@ -0,0 +1,356 @@
+ * Cache Prefixes
+ */
+
+export const memoryCachePrefixCommand = 'PF_COMMAND';
+export const memoryCachePrefixVersion = 'PF_VERSION';
+export const memoryCachePrefixCategory = 'PF_CATEGORY';
+export const memoryCachePrefixChannelDefaultVersion = 'PF_CHANNEL_VERSION';
```
```suggestion
export enum MemoryCachePrefix {
    COMMAND = 'PF_COMMAND',
    VERSION = 'PR_VERSION',
    CATEGORY = 'PR_CATEGORY',
    CHANNEL_DEFAULT_VERSION = 'PF_CHANNEL_VERSION',
}
```
I would use an enum here to better group the static exports together. Imo it also helps with readability. 

Apart from that, global static constants are usually written in screaming snake case. I only noticed this when I wondering where those variables were defined while reviewing another file. I'd change that for readability.

pdellaert replied:
Awesome suggestion, thanks!

ExampleWasTaken commented on src/lib/cache/cacheManager.ts:
```diff
@@ -0,0 +1,356 @@
+    for (const alias of aliases) {
+        // eslint-disable-next-line no-await-in-loop
+        await inMemoryCache.del(`${memoryCachePrefixCommand}:${alias.toLowerCase()}`);
+    }
```
```suggestion
    await Promise.all(aliases.map((alias) => inMemoryCache.del(`${memoryCachePrefixCommand}:${alias.toLowerCase()}`)));
```
This gets rid of your hated `no-await-in-loop` rule violation ;p

ExampleWasTaken commented on src/lib/cache/cacheManager.ts:
```diff
@@ -0,0 +1,356 @@
+    for (const alias of aliases) {
+        // eslint-disable-next-line no-await-in-loop
+        await inMemoryCache.set(`${memoryCachePrefixCommand}:${alias.toLowerCase()}`, command.toObject());
+    }
```
```suggestion
    await Promise.all(aliases.map((alias) => inMemoryCache.set(`${memoryCachePrefixCommand}:${alias.toLowerCase()}`, command.toObject())));
```
Satisfying ESLint and getting more efficient at the same time :)

ExampleWasTaken commented on src/lib/cache/cacheManager.ts:
```diff
@@ -0,0 +1,356 @@
+    const inMemoryCache = getInMemoryCache();
+    if (!conn || !inMemoryCache) return;
+
+    const PrefixCommands = await PrefixCommand.find();
```
```suggestion
    const prefixCommands = await PrefixCommand.find();
```
Probably just a typo?

ExampleWasTaken commented on src/lib/cache/cacheManager.ts:
```diff
@@ -0,0 +1,356 @@
+    for (const command of PrefixCommands) {
+        // eslint-disable-next-line no-await-in-loop
+        await loadSinglePrefixCommandToCache(command);
+    }
```
```suggestion
    await Promise.all(PrefixCommands.map((command) => loadSinglePrefixCommandToCache(command)));
```
`Promise.all()` is cool :D

ExampleWasTaken commented on src/lib/cache/cacheManager.ts:
```diff
@@ -0,0 +1,356 @@
+            // Step 3.b: If not found, remove from cache
+            if (!found) {
+                Logger.debug(`Removing command or alias ${checkCommand} from cache`);
+                // eslint-disable-next-line no-await-in-loop
+                await inMemoryCache.del(key);
```
I think readability would suffer significantly if this was done differently so `await` is fine here imo.

ExampleWasTaken commented on src/lib/cache/cacheManager.ts:
```diff
@@ -0,0 +1,356 @@
+    for (const dbCommand of prefixCommands) {
+        // eslint-disable-next-line no-await-in-loop
+        await loadSinglePrefixCommandToCache(dbCommand);
+    }
```
```suggestion
    await Promise.all(prefixCommands.map((dbCommand) => loadSinglePrefixCommandToCache(dbCommand)));
```
🙃

ExampleWasTaken commented on src/lib/cache/cacheManager.ts:
```diff
@@ -0,0 +1,356 @@
+    const inMemoryCache = getInMemoryCache();
+    if (!conn || !inMemoryCache) return;
+
+    const PrefixCommandVersions = await PrefixCommandVersion.find();
```
```suggestion
    const prefixCommandVersions = await PrefixCommandVersion.find();
```
I reckon that's another typo? :)

ExampleWasTaken commented on src/lib/cache/cacheManager.ts:
```diff
@@ -0,0 +1,356 @@
+    for (const version of PrefixCommandVersions) {
+        // eslint-disable-next-line no-await-in-loop
+        await loadSinglePrefixCommandVersionToCache(version);
+    }
```
```suggestion
    await Promise.all(PrefixCommandVersions.map((version) => loadSinglePrefixCommandVersionToCache(version)));
```
I'm running out of ideas of what to write here 😂

ExampleWasTaken commented on src/lib/cache/cacheManager.ts:
```diff
@@ -0,0 +1,356 @@
+    const inMemoryCache = getInMemoryCache();
+    if (!conn || !inMemoryCache) return;
+
+    // Step 1: Get all versions from the database
```
I really like these comments! They really help understand what the code does.

ExampleWasTaken commented on src/lib/cache/cacheManager.ts:
```diff
@@ -0,0 +1,356 @@
+    for (const dbVersion of prefixCommandVersions) {
+        // eslint-disable-next-line no-await-in-loop
+        await loadSinglePrefixCommandVersionToCache(dbVersion);
+    }
```
```suggestion
    await Promise.all(prefixCommandVersions.map((dbVersion) => loadSinglePrefixCommandVersionToCache(dbVersion)));
```
🤓

ExampleWasTaken commented on src/lib/cache/cacheManager.ts:
```diff
@@ -0,0 +1,356 @@
+    const inMemoryCache = getInMemoryCache();
+    if (!conn || !inMemoryCache) return;
+
+    const PrefixCommandCategories = await PrefixCommandCategory.find();
```
```suggestion
    const prefixCommandCategories = await PrefixCommandCategory.find();
```
Maybe these aren't typos after all? I would use camel case for them as they're normal variables but I'm also looking forward to hear what the idea behind using pascal case is!

pdellaert replied:
They are just one typo perpetuating with copy/paste stuff... Stupid on my end :)

ExampleWasTaken commented on src/lib/cache/cacheManager.ts:
```diff
@@ -0,0 +1,356 @@
+    for (const category of PrefixCommandCategories) {
+        // eslint-disable-next-line no-await-in-loop
+        await loadSinglePrefixCommandCategoryToCache(category);
+    }
```
```suggestion
    await Promise.all(PrefixCommandCategories.map((category) => loadSinglePrefixCommandCategoryToCache(category)));
```

Nerd level increasing... xD

ExampleWasTaken commented on src/lib/cache/cacheManager.ts:
```diff
@@ -0,0 +1,356 @@
+    for (const dbCategory of prefixCommandCategories) {
+        // eslint-disable-next-line no-await-in-loop
+        await loadSinglePrefixCommandCategoryToCache(dbCategory);
+    }
```
```suggestion
    await Promise.all(prefixCommandCategories.map((dbCategory) => loadSinglePrefixCommandCategoryToCache(dbCategory)));
```
Awaiting *all* the promises...

ExampleWasTaken commented on src/lib/cache/cacheManager.ts:
```diff
@@ -0,0 +1,356 @@
+    for (const defaultVersion of PrefixCommandChannelDefaultVersions) {
+        // eslint-disable-next-line no-await-in-loop
+        await loadSinglePrefixCommandChannelDefaultVersionToCache(defaultVersion);
+    }
```
```suggestion
    await Promise.all(
        PrefixCommandChannelDefaultVersions
            .map((defaultVersion) => loadSinglePrefixCommandChannelDefaultVersionToCache(defaultVersion)),
    );
```
Can't wait till prettier does the formtting for us...

ExampleWasTaken commented on src/lib/cache/cacheManager.ts:
```diff
@@ -0,0 +1,356 @@
+    for (const dbChannelDefaultVersion of prefixCommandChannelDefaultVersions) {
+        // eslint-disable-next-line no-await-in-loop
+        await loadSinglePrefixCommandChannelDefaultVersionToCache(dbChannelDefaultVersion);
+    }
```
```suggestion
    await Promise.all(
        prefixCommandChannelDefaultVersions
            .map(
                (dbChannelDefaultVersion) => loadSinglePrefixCommandChannelDefaultVersionToCache(dbChannelDefaultVersion),
            ),
    );
```
Even longer lmao. But I'm always a fan of descriptive names so I rather take that than guessing what a function may do 👍.

ExampleWasTaken commented on src/lib/schedulerJobs/refreshInMemoryCache.ts:
```diff
@@ -0,0 +1,36 @@
+
+    const start = new Date().getTime();
+    try {
+        await refreshAllPrefixCommandVersionsCache();
+        await refreshAllPrefixCommandCategoriesCache();
+        await refreshAllPrefixCommandsCache();
+        await refreshAllPrefixCommandChannelDefaultVersionsCache();
```
These could potentially be refreshed concurrently using `Promise.all()`, but it is unclear to me if, and if so, how versions and default channel version are dependent on each other. It may be simpler to avoid race conditions and update them sequentially.

pdellaert replied:
There is a risk for a race condition, I think. But it doesn't matter in what order caches are updated... The scenario is this:
Situation: DB Has a new version, and content for the new version for several commands. MEMCACHE does not have the version or content.

1. Cache update is triggered
2. New version is updated first. Command is lagging behind
3. User triggers command for that version (somehow they know it exists)
4. Handler picks up Version exists - But command has no content for it because command cache hasn't been loaded yet.
>> This should result in the handler to fall back on generic or not show anything

There's similar situations that can happen at different times, I guess. But I think most things should be captured somewhere. 

My biggest worry is what happens if someone updates a command, content, ... at the same time as the cache update runs. _WORST_ case:
* T0: Cache refresh DB Query pulls data of command content with `OLDCONTENT`
* T0+1MS: User Saves `NEWCONTENT`, DB is updated, Caches is updated
* T0+10MS: Cache refresh updates cache back with `OLDCONTENT`

BUT, the thing here is, only the cache is poisoned... There's two possibilities:
* User doesn't notice, next cache refresh pulls `NEWCONTENT` from DB, and updates cache
* User does notice, does the change again, all good. Or does a manual cache refresh to pull it.

This is the important factor of the cache, it's one direction: DB ==> CACHE. _NEVER_ CACHE ==> DB...

pdellaert replied:
BTW, cache refreshes take <0.5ms... While there is some risk for a missed cache update, I feel like it'll be manageable...

ExampleWasTaken replied:
Absolutely agree! It's totally not worth the effort.

ExampleWasTaken reviewed (commented):
This is outstanding work! The way this is thought through and structured is more than impressive!

I've said a few things about it on Discord already so here just the most important stuff:

Currently, all commands that list something (version, categories, commands) will fail when there are more than 25 items to list. This is due to Discord's limitation of max 25 fields per embed. I think these commands migth need a paginated embed.

Absolutely awesome stuff!

pdellaert committed:
Adding mod channel check suggestion

Co-authored-by: ExampleWasTaken <58574351+ExampleWasTaken@users.noreply.github.com>

pdellaert committed:
Language fix

Co-authored-by: ExampleWasTaken <58574351+ExampleWasTaken@users.noreply.github.com>

pdellaert committed:
Adding mod channel check suggestion

Co-authored-by: ExampleWasTaken <58574351+ExampleWasTaken@users.noreply.github.com>

pdellaert committed:
Language fix

Co-authored-by: ExampleWasTaken <58574351+ExampleWasTaken@users.noreply.github.com>

pdellaert committed:
Adding mod channel check suggestion

Co-authored-by: ExampleWasTaken <58574351+ExampleWasTaken@users.noreply.github.com>

pdellaert committed:
Cleanup unnecessary lint comment

Co-authored-by: ExampleWasTaken <58574351+ExampleWasTaken@users.noreply.github.com>

pdellaert committed:
Adding mod channel check suggestion

Co-authored-by: ExampleWasTaken <58574351+ExampleWasTaken@users.noreply.github.com>

pdellaert committed:
Language fix

Co-authored-by: ExampleWasTaken <58574351+ExampleWasTaken@users.noreply.github.com>

pdellaert committed:
Removing unnecessary lint comments

Co-authored-by: ExampleWasTaken <58574351+ExampleWasTaken@users.noreply.github.com>

pdellaert committed:
Adding mod channel check suggestion

Co-authored-by: ExampleWasTaken <58574351+ExampleWasTaken@users.noreply.github.com>

pdellaert committed:
Language fix

Co-authored-by: ExampleWasTaken <58574351+ExampleWasTaken@users.noreply.github.com>

pdellaert committed:
Adding mod channel check suggestion

Co-authored-by: ExampleWasTaken <58574351+ExampleWasTaken@users.noreply.github.com>

pdellaert committed:
Adding mod channel check suggestion

Co-authored-by: ExampleWasTaken <58574351+ExampleWasTaken@users.noreply.github.com>

pdellaert committed:
Language fix

Co-authored-by: ExampleWasTaken <58574351+ExampleWasTaken@users.noreply.github.com>

pdellaert committed:
Adding mod channel check suggestion

Co-authored-by: ExampleWasTaken <58574351+ExampleWasTaken@users.noreply.github.com>

pdellaert committed:
Language fix

Co-authored-by: ExampleWasTaken <58574351+ExampleWasTaken@users.noreply.github.com>

pdellaert committed:
Adding mod channel check suggestion

Co-authored-by: ExampleWasTaken <58574351+ExampleWasTaken@users.noreply.github.com>

pdellaert committed:
Adding language suggestion

Co-authored-by: ExampleWasTaken <58574351+ExampleWasTaken@users.noreply.github.com>

pdellaert committed:
Adding mod channel check suggestion

Co-authored-by: ExampleWasTaken <58574351+ExampleWasTaken@users.noreply.github.com>

pdellaert committed:
Adding mod channel check suggestion

Co-authored-by: ExampleWasTaken <58574351+ExampleWasTaken@users.noreply.github.com>

pdellaert committed:
Adding mod channel check suggestion

Co-authored-by: ExampleWasTaken <58574351+ExampleWasTaken@users.noreply.github.com>

pdellaert committed:
Adding mod channel check suggestion

Co-authored-by: ExampleWasTaken <58574351+ExampleWasTaken@users.noreply.github.com>

pdellaert committed:
Adding mod channel check suggestion

Co-authored-by: ExampleWasTaken <58574351+ExampleWasTaken@users.noreply.github.com>

pdellaert committed:
Updating Mod log checks and clean up some imports

pdellaert committed:
Increasing max lengths to be safe

pdellaert committed:
Fix bug on length, max 4000. Fix bug on filter

pdellaert committed:
Improving type checks

pdellaert committed:
Removing useless code

pdellaert committed:
Moving to enum for prefixes

pdellaert committed:
Fixing typos

pdellaert committed:
Migrating all possible loops to Promise.all

pdellaert committed:
Better to hard type

pdellaert committed:
Optimizing cache updates in parallel

pdellaert committed:
Limiting list outputs to 20 max

pdellaert committed:
Correcting changelog
